### PR TITLE
Add rcorrector

### PR DESCRIPTION
**Pull request by bimbam23**
this is a new tool wrapper to correct paired end reads. 
Wrapped tool: Rcorrector

### CODE REVIEW

**Comment by bimbam23** [comment]
hmm now I need a hint.

**Comment by bgruening** [comment]
We lint your python files as well :)
These are the errors:
```
./tools/rcorrector/FilterUncorrectabledPEfastq.py:22:1: F401 'itertools.izip' imported but unused
./tools/rcorrector/FilterUncorrectabledPEfastq.py:77:17: E265 block comment should start with '# '
./tools/rcorrector/FilterUncorrectabledPEfastq.py:78:17: E265 block comment should start with '# '
```

**Comment by bimbam23** [comment]
hmm can I have one of theses nice hints again?

**Comment by bgruening** [comment]
This is the error command:
```
 /home/travis/build/galaxyproject/tools-iuc/tools/rcorrector/rcorrector -k 23 -t 1 -maxcorK 4 -wk 0.95 -od output_file_directory  -p input_1.fq input_2.fq -c tmp_2c7a3e21a1028098f2b43546f6e2d98d.jf_dump
````
and this is the error:
```
  Failed at stage 3.
```

File: tools/rcorrector/FilterUncorrectabledPEfastq.xml [hunk 1]
```diff
@@ -0,0 +1,48 @@
+<tool id="FilterUncorrectabledPEfastq" name="Filter uncorrect PE reads from Rcorrector" version="0.1.0">
+    <description>for paired-end reads only</description>
+    <requirements>
+        <requirement type="package" version="2.7">python</requirement>
+    </requirements>
+    <command detect_errors="exit_code"><![CDATA[
+        ln -s '$input1' input_1.fq && ln -s '$input2' input_2.fq &&
+        python $__tool_directory__/FilterUncorrectabledPEfastq.py
```
Comment (bgruening) [inline tools/rcorrector/FilterUncorrectabledPEfastq.xml]: quote the python file as well.

File: tools/rcorrector/FilterUncorrectabledPEfastq.xml [hunk 2]
```diff
@@ -0,0 +1,48 @@
+<tool id="FilterUncorrectabledPEfastq" name="Filter uncorrect PE reads from Rcorrector" version="0.1.0">
+    <description>for paired-end reads only</description>
+    <requirements>
+        <requirement type="package" version="2.7">python</requirement>
+    </requirements>
+    <command detect_errors="exit_code"><![CDATA[
+        ln -s '$input1' input_1.fq && ln -s '$input2' input_2.fq &&
+        python $__tool_directory__/FilterUncorrectabledPEfastq.py 
+        -1 input_1.fq -2 input_2.fq -o fixed 2>&1 > rmunfixable.log && cat rmunfixable.log
```
Comment (bgruening) [inline tools/rcorrector/FilterUncorrectabledPEfastq.xml]: I'm confused here, the python script is creating a file with the same name and you are overwriting it?

File: tools/rcorrector/FilterUncorrectabledPEfastq.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    unfix_count = 0
+    r1_stream, r2_stream = get_input_streams(opts.leftreads, opts.rightreads)
+    with r1_stream as f1, r2_stream as f2:
+        R1 = grouper(f1, 4)
+        R2 = grouper(f2, 4)
+        counter = 0
+        for entry in R1:
+            counter += 1
+            if counter % 100000 == 0:
+                print "%s reads processed" % counter
```
Comment (bgruening) [inline tools/rcorrector/FilterUncorrectabledPEfastq.py]: Would be nice if you could port this script also to python3.

File: tools/rcorrector/FilterUncorrectabledPEfastq.xml [hunk 4]
```diff
@@ -0,0 +1,48 @@
+<tool id="FilterUncorrectabledPEfastq" name="Filter uncorrect PE reads from Rcorrector" version="0.1.0">
+    <description>for paired-end reads only</description>
```
Comment (bgruening) [inline tools/rcorrector/FilterUncorrectabledPEfastq.xml]: @bimbam23 this tools looks very very special for rcorrector, can we not put this python script directly in the rcorrector tool?

File: tools/rcorrector/rcorrector.xml [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+            -s input_1.fq
+        #end if
+        #if $library.lib == "paired":
+            -1 input_1.fq -2 input_2.fq
+        #end if
+        -k '$advanced.kmers' -t \${GALAXY_SLOTS:-4} -maxcorK '$advanced.maxcorK' -wk '$advanced.wk' -ek '$advanced.ek' -od output_file_directory 
+        
+        #if $library.lib == "paired":
+            #if $library.filter:
+                && python $__tool_directory__/FilterUncorrectabledPEfastq.py
```
Comment (bgruening) [inline tools/rcorrector/rcorrector.xml]: Mh, the tool is here again? Ok, I guess I need sleep. Do we need the extra Galaxy tool for FilterUncorrectabledPEfastq then?

File: tools/rcorrector/rcorrector.xml [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+            @article{Rcorrector,
+            author = {Song, Li and Florea, Liliana},
+            title = {Rcorrector: efficient and accurate error correction for Illumina RNA-seq reads},
+            journal = {GigaScience},
+            volume = {4},
+            number = {1},
+            pages = {1-8},
+            year = {2015},
+            doi = {10.1186/s13742-015-0089-y},
+            URL = {http://dx.doi.org/10.1186/s13742-015-0089-y},
```
Comment (bgruening) [inline tools/rcorrector/rcorrector.xml]: can you use the DOI as citation type here, please?

File: tools/rcorrector/run_rcorrector.pl [hunk 7]
```diff
@@ -0,0 +1,273 @@
+#!/usr/bin/env perl
```
Comment (bgruening) [inline tools/rcorrector/run_rcorrector.pl]: I'm unsure, but we could create a conda package for this

File: tools/rcorrector/FilterUncorrectabledPEfastq.xml (hunk 2 above)
Comment (bimbam23) [reply tools/rcorrector/FilterUncorrectabledPEfastq.xml]: The idea is to catch the log output but I guess it work also without saving it into an extra file. :-)

File: tools/rcorrector/FilterUncorrectabledPEfastq.py (hunk 3 above)
Comment (bimbam23) [reply tools/rcorrector/FilterUncorrectabledPEfastq.py]: hmm the tool is not mine I took it from https://github.com/harvardinformatics/TranscriptomeAssemblyTools/blob/master/FilterUncorrectabledPEfastq.py

File: tools/rcorrector/rcorrector.xml (hunk 5 above)
Comment (bimbam23) [reply tools/rcorrector/rcorrector.xml]: I did not know include a stand-alone tool inside another tool.

File: tools/rcorrector/run_rcorrector.pl (hunk 7 above)
Comment (bimbam23) [reply tools/rcorrector/run_rcorrector.pl]: I put this into the rcorretor package but somehow it cannot be accessed

File: tools/rcorrector/FilterUncorrectabledPEfastq.xml (hunk 4 above)
Comment (bimbam23) [reply tools/rcorrector/FilterUncorrectabledPEfastq.xml]: its not a script from the original authors ... I would not use it in any cases and thereby leave it outside the rcorrector tool

File: tools/rcorrector/run_rcorrector.pl (hunk 7 above)
Comment (bgruening) [reply tools/rcorrector/run_rcorrector.pl]: Than we should fix this conda package :)
https://github.com/bioconda/bioconda-recipes/pull/8858

File: tools/rcorrector/FilterUncorrectabledPEfastq.xml (hunk 4 above)
Comment (bgruening) [reply tools/rcorrector/FilterUncorrectabledPEfastq.xml]: The author does not really matter. You are maintaining the tools here :)
My point if, if this tool is really only useful for rcorrector output, we also can put this as post-precessing in the tool directly as we do not gain much my separating them. Instead we get more tools to maintain/update/teach and pollute the toolbox. Just my opinion.

Comment (bimbam23) [reply tools/rcorrector/FilterUncorrectabledPEfastq.xml]: its true the tool is only to be used with rcorrector, so we could add it to the conda package.

File: tools/rcorrector/run_rcorrector.pl (hunk 7 above)
Comment (bimbam23) [reply tools/rcorrector/run_rcorrector.pl]: here is the last line of the PR of bulid.sh
`cp -R rcorrector run_rcorrector.pl $PREFIX/bin`
I think it should work?  I copying it into the correct path .

so with `perl run_rcorrector.pl` it should work?

Comment (bgruening) [reply tools/rcorrector/run_rcorrector.pl]: you did not set the `+x` flag on the script.

Comment (bimbam23) [reply tools/rcorrector/run_rcorrector.pl]: jep I just had a look a the link :-)

Comment (bimbam23) [reply tools/rcorrector/run_rcorrector.pl]: thanks for fixing

**Comment by bimbam23** [comment]
´Can't open perl script "run_rcorrector.pl": No such file or directory´
it looks like conda is still not providing the tool?

**Comment by bimbam23** [comment]
@bgruening do you have an idea for the perl script?

**Comment by bimbam23** [comment]
Should I change it back... Than it's working at least.

**Comment by bimbam23** [comment]
hmm is there anything I can do to get this fixed? Are perl scripts in bioconda still on the blacklist?

**Comment by bimbam23** [comment]
@nsoranzo do you have any idea?

File: tools/rcorrector/rcorrector.xml [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+    </requirements>
+    <command detect_errors="exit_code"><![CDATA[
+        #if $library.lib == "single":
+            ln -s '$library.input1' input_1.fq &&
+        #end if
+        #if $library.lib == "paired":
+            ln -s '$library.input1' input_1.fq && ln -s '$library.input2' input_2.fq &&
+        #end if
+
+        perl run_rcorrector.pl
```
Comment (nsoranzo) [inline tools/rcorrector/rcorrector.xml]: Remove `perl`, the script is installed from bioconda in `$PATH`.

File: tools/rcorrector/rcorrector.xml [hunk 9]
```diff
@@ -0,0 +1,122 @@
+<tool id="rcorrector" name="RNA-seq Rcorrector" version="0.1.0">
+	<description>a kmer-based error correction method for RNA-seq data</description>
+    <requirements>
+        <requirement type="package" version="5.22">perl</requirement>
+        <requirement type="package" version="1.0.2-1">rcorrector</requirement>
```
Comment (nsoranzo) [inline tools/rcorrector/rcorrector.xml]: You can update to perl 5.26.2 and rcorrector 1.0.3 .

File: tools/rcorrector/rcorrector.xml (hunk 8 above)
Comment (nsoranzo) [reply tools/rcorrector/rcorrector.xml]: This https://github.com/bioconda/bioconda-recipes/pull/8858/files#r214879183 needs to be fixed before `run_rcorrector.pl` works properly.

Comment (bimbam23) [reply tools/rcorrector/rcorrector.xml]: that exactly has to be changed here?

Comment (nsoranzo) [reply tools/rcorrector/rcorrector.xml]: Remove `perl`, from:
```
perl run_rcorrector.pl
```
to just:
```
run_rcorrector.pl
```

Comment (nsoranzo) [reply tools/rcorrector/rcorrector.xml]: The issue I mentioned [above](https://github.com/galaxyproject/tools-iuc/pull/1861#discussion_r214880359) has been fixed.

**Comment by nsoranzo** [comment]
@bimbam23 There was some strange interaction between Travis and Planemo resulting in many unrelated tools being tested (see https://travis-ci.org/galaxyproject/tools-iuc/builds/424299529 ), which I was not able to reproduce locally.
I've therefore rebased this pull request, hopefully this will fix it.

**Comment by nsoranzo** [comment]
Green!

**Comment by bimbam23** [comment]
Nice work!!

File: tools/rcorrector/.shed.yml [hunk 10]
```diff
@@ -0,0 +1,10 @@
+owner: iuc
+name: rcorrector
+description: Rcorrector (RNA-seq error CORRECTOR) is a kmer-based error correction method for RNA-seq data.
+homepage_url: https://github.com/mourisl/Rcorrector
+long_description: |
```
[the remark this answers is not in the document]
Comment (bimbam23) [reply tools/rcorrector/.shed.yml]: hmm so the second small tool can only be used after rcorrector ... so I guess we need a suite?

**Comment by bgruening** [comment]
Thanks @bimbam23

**Pull request merged**